tests/test_trajectory.py

```diff
@@ -0,0 +1,305 @@
+    })
+    deviation_survey_file_path = os.path.join(model.epc_directory, 'deviation_survey.csv')
+    source_dataframe.to_csv(deviation_survey_file_path)
+
+    for well_name in well_names:
+        if well_name is None:
+            # --------- Act ----------
+            # Create a trajectory from the csv file
+            with pytest.raises(Exception) as excinfo:
+                trajectory_from_ascii = resqpy.well.Trajectory(parent_model = model,
+                                                               deviation_survey_file = deviation_survey_file_path,
+                                                               length_uom = 'm',
+                                                               md_datum = datum)
+
+            # -------- Assert ---------
+            assert "attempt to set trajectory for unidentified well from ascii file holding data for multiple wells" in str(
+                excinfo.value)
+        else:
+            # --------- Act ----------
+            # Create a trajectory from the csv file
+            trajectory_from_ascii = resqpy.well.Trajectory(parent_model = model,
+                                                           deviation_survey_file = deviation_survey_file_path,
+                                                           well_name = well_name,
+                                                           length_uom = 'm',
+                                                           md_datum = datum)
+            # -------- Assert ---------
+            assert trajectory_from_ascii is not None
+
+
+def test_set_tangents(example_model_and_crs):
+    # --------- Arrange ----------
+    model, crs = example_model_and_crs
+    elevation = 100
+    # Create a measured depth datum
+    datum = resqpy.well.MdDatum(parent_model = model,
+                                crs_uuid = crs.uuid,
+                                location = (0, 0, -elevation),
+                                md_reference = 'kelly bushing')
+    mds = np.array([300, 310, 330, 340])
+    zs = mds - elevation
+    well_name = 'Coconut'
+    source_dataframe = pd.DataFrame({
+        'MD': mds,
+        'X': [1, 2, 3, 4],
+        'Y': [1, 2, 3, 4],
+        'Z': zs,
+        'WELL': ['Coconut', 'Coconut', 'Coconut', 'Coconut']
+    })
+
+    # Create a trajectory from dataframe
+    trajectory = resqpy.well.Trajectory(parent_model = model,
+                                        data_frame = source_dataframe,
+                                        well_name = well_name,
+                                        md_datum = datum,
+                                        length_uom = 'm')
+
+    # --------- Act ----------
+    # Calculate tangent vectors based on control points
+    trajectory.set_tangents()
+
+    # --------- Assert ----------
+    assert trajectory.tangent_vectors is not None
+    assert trajectory.tangent_vectors.shape == (trajectory.knot_count, 3)
+
```

Comment:
You could use some vector_utilities to perform some further basic checks on the tangents, eg.:
naive length of each tangent is close to 1
with the xyz data you are using, I would expect tangent[1] and tangent[2] to be more-or-less the same

Reply:
Considering I used the most basic, x = y line for the X and Y coordinates, it should have been obvious that the azimuths of the tangent vectors would all be 45 :D.